test/e2e/compact_test.go

```diff
@@ -0,0 +1,138 @@
+	testutil.Ok(t, err)
+	testutil.Ok(t, objstore.UploadDir(ctx, l, bkt, path.Join(dir, id1.String()), id1.String()))
+
+	id2, err := e2eutil.CreateBlockWithBlockDelay(ctx, dir, series, 10, timestamp.FromTime(now), timestamp.FromTime(now.Add(2*time.Hour)), 30*time.Minute, extLset2, 0)
+	testutil.Ok(t, err)
+	testutil.Ok(t, objstore.UploadDir(ctx, l, bkt, path.Join(dir, id2.String()), id2.String()))
+
+	id3, err := e2eutil.CreateBlockWithBlockDelay(ctx, dir, series, 10, timestamp.FromTime(now), timestamp.FromTime(now.Add(2*time.Hour)), 30*time.Minute, extLset3, 0)
+	testutil.Ok(t, err)
+	testutil.Ok(t, objstore.UploadDir(ctx, l, bkt, path.Join(dir, id3.String()), id3.String()))
+
+	cmpt, err := e2ethanos.NewCompactor(s.SharedDir(), "1", client.BucketConfig{
+		Type: client.S3,
+		Config: s3.Config{
+			Bucket:    bucket,
+			AccessKey: e2edb.MinioAccessKey,
+			SecretKey: e2edb.MinioSecretKey,
+			Endpoint:  m.NetworkHTTPEndpoint(),
+			Insecure:  true,
+		},
+	},
+		"--deduplication.replica-label=replica",
+		"--deduplication.replica-label=rule_replica",
+	)
+
+	testutil.Ok(t, err)
+	testutil.Ok(t, s.StartAndWaitReady(cmpt))
+
+	testutil.Ok(t, cmpt.WaitSumMetrics(e2e.Equals(3), "thanos_blocks_meta_synced"))
+	testutil.Ok(t, cmpt.WaitSumMetrics(e2e.Equals(0), "thanos_blocks_meta_sync_failures_total"))
+	testutil.Ok(t, cmpt.WaitSumMetrics(e2e.Equals(3), "thanos_blocks_meta_modified"))
+
+	str, err := e2ethanos.NewStoreGW(s.SharedDir(), "1", client.BucketConfig{
+		Type: client.S3,
+		Config: s3.Config{
+			Bucket:    bucket,
+			AccessKey: e2edb.MinioAccessKey,
+			SecretKey: e2edb.MinioSecretKey,
+			Endpoint:  m.NetworkHTTPEndpoint(),
+			Insecure:  true,
+		},
+	})
+	testutil.Ok(t, err)
+	testutil.Ok(t, s.StartAndWaitReady(str))
+
+	testutil.Ok(t, str.WaitSumMetrics(e2e.Equals(1), "thanos_blocks_meta_synced"))
+	testutil.Ok(t, str.WaitSumMetrics(e2e.Equals(0), "thanos_blocks_meta_sync_failures_total"))
+
+	q, err := e2ethanos.NewQuerier(s.SharedDir(), "1", []string{str.GRPCNetworkEndpoint()}, nil)
+	testutil.Ok(t, err)
+	testutil.Ok(t, s.StartAndWaitReady(q))
+
+	ctx, cancel = context.WithTimeout(context.Background(), 3*time.Minute)
+	defer cancel()
+
+	t.Run("query works", func(t *testing.T) {
+		queryAndAssert(t, ctx, q.HTTPEndpoint(), "{a=\"1\"}",
+			promclient.QueryOptions{
+				Deduplicate: false, // This should be false, so that we can be sure deduplication was offline.
+			},
+			[]model.Metric{
+				{
+					"a":    "1",
+					"b":    "2",
+					"ext1": "value1",
+				},
+			},
+		)
+
+		// Make sure only necessary amount of blocks fetched from store, to observe affects of offline deduplication.
+		testutil.Ok(t, str.WaitSumMetrics(e2e.Equals(3), "thanos_bucket_store_series_data_touched"))
+		testutil.Ok(t, str.WaitSumMetrics(e2e.Equals(3), "thanos_bucket_store_series_data_fetched"))
+		testutil.Ok(t, str.WaitSumMetrics(e2e.Equals(1), "thanos_bucket_store_series_blocks_queried"))
```

Comment:
Hm.. I think we need to actually download meta.json and assert.. WDYT?

Reply:
Assert on number os series, samples etc.

Reply:
Alright, makes sense I'm gonna do it.